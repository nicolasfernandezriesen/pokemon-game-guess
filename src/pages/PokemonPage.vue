<template>
  <TimerCount :timer="timer" :showTimer="showTimer"/>

  <div id="header">
    <h3>Correctos : <span style="color: greenyellow;">{{ countGood }}</span></h3>
    <h3>Fallados : <span style="color: red;">{{ countBad }}</span></h3>
  </div>

  <h1 v-if="!pokemon">Cargando...</h1>
  <div v-else>
    <h1>¿Quien es este pokémon?</h1>

    <PokemonPicture :pokemonId="pokemon.id" :showPokemon="showPokemon"/>
    <PokemonOptions :pokemons="pokemonArr" @selectedPokemon="checkAnswer"/> 

    <template v-if="showAnswer">
      <h2 class="fade-in">{{ message }}</h2> 
    </template>  

    <div v-if="countBad || countGood > 0">
      <button @click="resetCount">
        Reiniciar contadores
      </button> 
    </div> 
  </div>
  
</template>

<script>
import PokemonPicture from '@/components/PokemonPicture.vue';
import PokemonOptions from '@/components/PokemonOptions.vue';
import getPokemonsOptions from '@/helpers/getPokemonOptions';
import TimerCount from '@/components/TimerCount.vue';

export default {
    components :{
        PokemonPicture,
        PokemonOptions,
        TimerCount
    },
    data(){
      return{
        pokemonArr: [],
        pokemon: false,
        showPokemon: false,
        showAnswer: false,
        message: '',
        countGood: 0,
        countBad: 0,
        timer: 3,
        showTimer: false
      }
    },
    methods:{
      async mixPokemonArray(){
        this.pokemonArr = await getPokemonsOptions();
        
        let rndInt = Math.floor(Math.random() *4);
        this.pokemon = this.pokemonArr[rndInt];
      },
      checkAnswer(pokemonId){
        this.showPokemon = true;
        this.showAnswer = true;

        if(pokemonId == this.pokemon.id){
          this.message = 'Correcto adivinaste!!!!';
          this.countGood++;
        }else{
          this.message = `Ohh, fallaste, el pokemon era ${this.pokemon.name}`;
          this.countBad++;
        }
        this.showTimer = true;
        setTimeout(this.newGame, 3000);
      },
      newGame(){
        this.showPokemon = false;
        this.showAnswer = false;
        this.pokemon = false;
        this.pokemonArr = [];
        this.showTimer = false;
        this.timer = 3;
        
        this.mixPokemonArray();
      },
      resetCount(){
        this.countBad = 0;
        this.countGood = 0;
        
        this.newGame();
      }
    },
    mounted(){
      this.mixPokemonArray();
    }
}
</script>

<style>
  #header{
    align-items: center;
    justify-content: center;
    display: flex;
  }

  h3{
    padding: 5px;
  }
</style>