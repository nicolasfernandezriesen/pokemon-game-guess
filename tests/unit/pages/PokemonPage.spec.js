import { shallowMount, mount } from "@vue/test-utils";
import PokemonPage from "@/pages/PokemonPage";
import { pokemons } from "../mocks/pokemons.mock";

describe('PokemonPage', () => {
    let wrapper;
    let mixPokemonArraySpy;
    
    beforeEach(() => {
        mixPokemonArraySpy = jest.spyOn(PokemonPage.methods, 'mixPokemonArray');       
        wrapper = shallowMount(PokemonPage, {
            data() {
                return {
                    pokemonArr: pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: '',
                    countGood: 0,
                    countBad: 0
                }
            }
        })
    })

    afterEach(() => {
       jest.clearAllMocks(); 
    })

    test('it should match with the snapshot without pokemon', () =>{
        let wrapper = shallowMount(PokemonPage);

        expect(wrapper.html).toMatchSnapshot();
    })

    test('it should match with the snapshot without pokemon', () =>{
        expect(wrapper.html).toMatchSnapshot();
    })

    test('it should call mixPokemonArray', () => {
        expect(mixPokemonArraySpy).toHaveBeenCalled();    
    })

    test('it should show the component from PokemonPicture & PokemonOptions', () => {
        let picture = wrapper.find('pokemon-picture-stub');
        let options = wrapper.find('pokemon-options-stub');

        expect(picture.exists()).toBeTruthy();
        expect(options.exists()).toBeTruthy();  
  
        expect(picture.attributes('pokemonid')).toBe("1"); 
        expect(options.attributes('pokemons')).toBeTruthy();
    })

    test('it should show the msg for the bad & good answer', async() => {
        await wrapper.vm.checkAnswer(1);
        let goodChoice = wrapper.vm.message;
        
        await wrapper.vm.checkAnswer(5);
        let badChoice = wrapper.vm.message;

        expect(wrapper.find('h2')).toBeTruthy();
        expect(wrapper.vm.showPokemon).toBe(true);
        expect(goodChoice).toBe('Correcto adivinaste!!!!');
        expect(badChoice).toBe(`Ohh, fallaste, el pokemon era ${pokemons[0].name}`);
    })

    test('it should increased the good & bad count according to the answer', async() => {
        await wrapper.vm.checkAnswer(1);
        await wrapper.vm.checkAnswer(1);
        let countGood = wrapper.vm.countGood;

        await wrapper.vm.checkAnswer(5);
        let countBad = wrapper.vm.countBad;

        expect(countGood).toBe(2);
        expect(countBad).toBe(1);
    })
})