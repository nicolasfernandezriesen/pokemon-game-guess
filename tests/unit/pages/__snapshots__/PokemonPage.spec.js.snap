// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage it should match with the snapshot without pokemon 1`] = `[Function]`;

exports[`PokemonPage it should match with the snapshot without pokemon 2`] = `[Function]`;

exports[`PokemonPage it should match with the snapshot without pokemons 1`] = `[Function]`;

exports[`PokemonPage it should match with the snapshot without pokemons 2`] = `[Function]`;
